File: problem-solving/_lc-7-medium-reverse-integer/1-solution.py
"""
https://leetcode.com/problems/reverse-integer/

Complexity
----------
Time: O(logn)
Space: O(1)

NOTE: log base 10
"""

MAX = (1 << 31)
class Solution:
    def reverse(self, x: int) -> int:
        sign = 1
        if x < 0:
            x = -x
            sign = -1
        res = 0
        while x:
            res = (res * 10) + (x % 10)
            x //= 10
        res = sign * res
        if res < -MAX or res > MAX - 1:
            return 0
        return res

File: problem-solving/_lc-0122-medium-best-time-to-buy-and-sell-stock-ii/1-solution.go
/*
https://leetcode.com/problems/best-time-to-buy-and-sell-stock-ii/

Complexity
----------
Time: O(n)
Space: O(1)

NOTE
----
This is a peak valley problem, we have to buy and valleys and sell at peaks.
*/


func maxProfit(prices []int) int {
    buy, profit, n := 0, 0, len(prices)
    for i := 0; i < n; i++ {
        for ;i < n - 1 && prices[i] > prices[i + 1]; i++ {}
        buy = prices[i]

        for ;i < n - 1 && prices[i] < prices[i + 1]; i++ {}
        profit += prices[i] - buy
    }
    return profit
}
